### PR TITLE
RegistryServer: Report back errors from OnNodeAdded

### PR DESCRIPTION
**Pull request by avalluri**
Failures happened in RegistryListener.OnNodeAdded() are supposed to send back to
appropriate node, and the node registration should fail. This allows the node
driver to make this error visible to cluster admins via driver health check.

FIXES: #320

### CODE REVIEW

File: pkg/registryserver/registryserver.go [hunk 1]
```diff
@@ -115,7 +117,12 @@ func (rs *RegistryServer) RegisterController(ctx context.Context, req *registry.
 
 	if !found {
 		for l := range rs.listeners {
-			l.OnNodeAdded(ctx, node)
+			if err := l.OnNodeAdded(ctx, node); err != nil {
```
Comment (pohly) [inline pkg/registryserver/registryserver.go]: This is now racy again:
- RegisterController is called twice for the same node
- first call adds the node, fails in OnNodeAdded
- second call finds that the node is registered
- first call removes the node and returns an error
- second call returns success
- node thinks that it is registered but in reality it isn't

The correct solution is to lock the entire RegisterController until it either succeeds (including the OnNodeAdded) or fails.

Injecting foreign code via this callback mechanism is what makes this all so complicated. Can registering nodes and establishing a connection to them be done entirely inside this file?

Comment (avalluri) [reply pkg/registryserver/registryserver.go]: Second `RegisterController `call from same node waits till the first call completes. There is mutex lock based on `NodeId`,

Comment (pohly) [reply pkg/registryserver/registryserver.go]: Okay, that works.

Can you add explanations of how this locking works to the source code?

Comment (avalluri) [reply pkg/registryserver/registryserver.go]: > Can you add explanations of how this locking works to the source code?

Done

**Pull request merged**